File: imasm_module.py
"""The module format: a word in the twelve, and the payload each glyph carries.

The glyph is the opcode. It says which of the twelve axes an instruction is,
and that alone decides how the payload is read — a ∈ carries a condition and a
target, a ⊞ carries an ALU operation and its operands, a ⊤ carries the two
things compared. Structure and data stay separate the whole way down, which is
why the word prints as glyphs with nothing else in it while the module still
runs.

Line format, one instruction per line:

    GLYPH \t field \t field ...

Operand fields are normalised so the machine never parses assembly:

    r:rax                          a register
    i:0x10                         an immediate
    m:base:index:scale:disp:size   a memory reference
"""
import vox_x86 as capstone
from vox_x86 import X86_OP_IMM, X86_OP_MEM, X86_OP_REG

ENTRY, TERM, SPLIT, FUSE = "⊢", "⊣", "∈", "∋"
CALL, XFER, INDIRECT, COMMIT = ">", "<", "⊙", "◻"
LINK, TRUTH, CONSUME, ENGAGE = "⋈", "⊤", "⊥", "⊞"

_MOVE = {"mov", "movzx", "movsx", "movsxd", "movabs", "push", "pop", "xchg",
         "leave"}          # leave is slot movement: rsp ← rbp, then rbp popped
_INERT = {"nop", "endbr64", "endbr32"}   # never a commit, whatever they address
_TRUTH = {"cmp", "test"}
_TERMINAL = {"ret", "retf", "int3", "ud2", "hlt", "iret"}


def _operand(ins, op) -> str:
    if op.type == X86_OP_REG:
        return "r:" + ins.reg_name(op.reg)
    if op.type == X86_OP_IMM:
        return f"i:{op.imm:#x}" if op.imm >= 0 else f"i:-{-op.imm:#x}"
    if op.type == X86_OP_MEM:
        m = op.mem
        base = ins.reg_name(m.base) if m.base else ""
        index = ins.reg_name(m.index) if m.index else ""
        return f"m:{base}:{index}:{m.scale}:{m.disp:#x}:{op.size}"
    return "?:"


_PREFIX = ("notrack ", "lock ", "bnd ", "rep ", "repe ", "repne ", "data16 ")


def _mnemonic(ins) -> str:
    """Prefixes ride on the mnemonic in capstone's text. They change how an
    instruction is guarded, not which of the twelve it is, so they come off
    before classification — a `notrack jmp` is a jmp."""
    mn = ins.mnemonic
    for p in _PREFIX:
        if mn.startswith(p):
            mn = mn[len(p):]
    return mn


def encode(ins, is_merge: bool) -> list:
    """One instruction → its lines. A merge emits a bare ∋ first, because the
    fuse is a property of the address, not of the instruction sitting on it."""
    import vox
    lines = [FUSE] if is_merge else []
    mn, ops = _mnemonic(ins), ins.operands
    f = [_operand(ins, o) for o in ops]
    direct = bool(ops) and ops[0].type == X86_OP_IMM

    # The glyph is vox.classify's call, not a second opinion. Only the payload
    # is decided here, so the module text and the auditor's word cannot differ
    # on which of the twelve an instruction is.
    g = vox.classify(ins)

    if g == TERM:
        lines.append(f"{TERM}\t{mn}")
    elif g == INDIRECT and mn in ("syscall", "sysenter", "int"):
        # A syscall's real target is chosen by the value in a register (rax),
        # not written anywhere in the instruction — the same structural shape
        # as an indirect call/jmp, just to the kernel instead of the program's
        # own code. ⊙ INDIRECT is what that shape is; no thirteenth glyph
        # needed for "transfer to something outside what was disassembled."
        lines.append(f"{INDIRECT}\tsyscall")
    elif mn == "call":
        lines.append("\t".join([g, mn] + f))
    elif mn == "jmp":
        lines.append("\t".join([g, mn] + f))
    elif mn.startswith("j"):
        lines.append("\t".join([SPLIT, mn[1:]] + f))
    elif mn.startswith("set"):
        lines.append("\t".join([CONSUME, mn[3:], "set"] + f))
    elif mn.startswith("cmov"):
        lines.append("\t".join([CONSUME, mn[4:], "cmov"] + f))
    elif g == TRUTH:
        lines.append("\t".join([TRUTH, mn] + f))
    elif g == COMMIT:
        lines.append("\t".join([COMMIT, mn] + f))
    elif g == LINK:
        lines.append("\t".join([LINK, mn] + f))
    else:
        lines.append("\t".join([ENGAGE, mn] + f))
    return lines


def disassembler(path: str):
    """The decoder with operand detail on — the payload comes from structure,
    not from re-parsing assembly text. Recursive descent, function-grouped, the
    same walk the auditor uses — not a second, separate linear sweep."""
    import vox
    with open(path, "rb") as fh:
        magic = fh.read(4)
    parse = vox._elf_sections if magic == b"\x7fELF" else vox._pe_sections
    _mode, _secs, entry = parse(path)
    funcs = list(vox._native_functions(path))
    insns = [ins for _, func in funcs for ins in func]
    return insns, entry, funcs


def _plt_map(path: str):
    """(stub address -> external symbol name), ELF x86-64 only — see
    vox._plt_stub_map. PE import-table resolution isn't attempted; a PE's
    external calls stay unresolved, same as before."""
    import vox
    with open(path, "rb") as fh:
        magic = fh.read(4)
    if magic != b"\x7fELF":
        return {}
    mode, secs, _entry = vox._elf_sections(path)
    return vox._plt_stub_map(path, secs, mode)


def data_sections(path: str):
    """Initialised bytes the code will read: constants, tables, strings. A
    module with no data is not executable — a rip-relative load of a vector
    constant reads zeros and the program quietly computes the wrong answer."""
    import struct
    with open(path, "rb") as fh:
        raw = fh.read()
    if raw[:4] == b"\x7fELF":
        is64 = raw[4] == 2
        end = "<" if raw[5] == 1 else ">"
        off_shoff, fmt = (40, "QQQ") if is64 else (32, "III")
        e_shoff = struct.unpack_from(end + fmt, raw, 24)[2]
        e_shentsize, e_shnum = struct.unpack_from(
            end + "HH", raw, 58 if is64 else 46)
        out = []
        for k in range(e_shnum):
            o = e_shoff + k * e_shentsize
            if is64:
                _, sh_type, sh_flags, sh_addr, sh_off, sh_size = \
                    struct.unpack_from(end + "IIQQQQ", raw, o)
            else:
                _, sh_type, sh_flags, sh_addr, sh_off, sh_size = \
                    struct.unpack_from(end + "IIIIII", raw, o)
            # ALLOC, PROGBITS, not executable: the data the code reads
            if sh_type == 1 and sh_flags & 0x2 and not sh_flags & 0x4 and sh_size:
                out.append((sh_addr, raw[sh_off:sh_off + sh_size]))
        return out
    import pefile
    pe = pefile.PE(path, fast_load=True)
    base = pe.OPTIONAL_HEADER.ImageBase
    return [(base + s.VirtualAddress, s.get_data()) for s in pe.sections
            if not s.Characteristics & 0x20000000 and s.SizeOfRawData]



def _merges_of_function(insns) -> set:
    """Compute merge points (addresses with >=2 predecessors) for a single function.
    Uses the same prefix-stripping logic as _mnemonic to stay consistent with vox._merges_of."""
    from collections import Counter
    aset = {i.address for i in insns}
    succ = []
    for idx, ins in enumerate(insns):
        mn = _mnemonic(ins)
        if not (mn == "jmp" or mn.startswith("ret")) \
                and idx + 1 < len(insns):
            succ.append(insns[idx + 1].address)
        if mn.startswith("j") and ins.operands \
                and ins.operands[0].type == X86_OP_IMM:
            if ins.operands[0].imm in aset:
                succ.append(ins.operands[0].imm)
    return {a for a, c in Counter(succ).items() if c >= 2}


def emit(path: str) -> str:
    """The whole binary as an executable IMASM module. Recursive descent,
    function by function; a function that is a known PLT stub (see
    vox._plt_stub_map) is emitted as a single external-call line instead of
    the endbr64/indirect-jump-through-an-unloaded-GOT-slot it really holds —
    the machine never sees a binary, so that jump has nothing real to read.
    """
    from collections import Counter
    insns, entry, funcs = disassembler(path)
    stubs = _plt_map(path)

    out = [f"; ⊙ {path}", f"; entry 0x{entry:x}"]
    for addr, blob in data_sections(path):
        out.append(f"={addr:#x}\t{blob.hex()}")
    for start, func in funcs:
        if start in stubs:
            out.append(f"@0x{start:x}")
            out.append(f"{INDIRECT}\texternal\t{stubs[start]}")
            continue
        # Compute merges PER FUNCTION, matching vox._merges_of
        func_merges = _merges_of_function(func)
        for ins in func:
            out.append(f"@0x{ins.address:x}")
            out.extend(encode(ins, ins.address in func_merges))
    return "\n".join(out) + "\n"

File: src/imasm_module.rs
//! The module format: a word in the twelve, and the executable payload each
//! glyph carries. Ported from imasm_module.py. The glyph is the opcode (which of
//! the twelve axes the instruction is); the payload is normalised operands the
//! machine reads without ever parsing assembly.
//!
//!     GLYPH \t field \t field ...      with  r:reg  i:imm  m:base:index:scale:disp:size

use alloc::string::{String, ToString};
use alloc::vec::Vec;
use alloc::collections::{BTreeMap, BTreeSet};
use alloc::format;

use crate::x86::{self, Op};
use crate::vox_decode;

const ENTRY: char = '⊢'; const TERM: char = '⊣'; const SPLIT: char = '∈'; const FUSE: char = '∋';
const CALL: char = '>'; const XFER: char = '<'; const INDIRECT: char = '⊙'; const COMMIT: char = '◻';
const LINK: char = '⋈'; const TRUTH: char = '⊤'; const CONSUME: char = '⊥'; const ENGAGE: char = '⊞';

fn is_move(mn: &str) -> bool {
    matches!(mn, "mov"|"movzx"|"movsx"|"movsxd"|"movabs"|"push"|"pop"|"xchg"|"leave")
}
fn is_truth(mn: &str) -> bool { matches!(mn, "cmp"|"test") }
fn is_terminal(mn: &str) -> bool { mn.starts_with("ret") || matches!(mn, "int3"|"ud2"|"hlt"|"iret"|"retf") }

/// The glyph — the same decision vox's classifier makes.
fn classify(i: &x86::Insn) -> char {
    let mn = i.mnemonic.as_str();
    if is_terminal(mn) { return TERM; }
    if mn == "call" { return if i.target.is_some() { CALL } else { INDIRECT }; }
    if mn == "jmp"  { return if i.target.is_some() { XFER } else { INDIRECT }; }
    if mn == "syscall" || mn == "sysenter" || mn == "int" { return INDIRECT; }
    if mn.starts_with('j') && mn != "jmp" { return SPLIT; }
    if mn.starts_with("set") { return CONSUME; }
    if mn.starts_with("cmov") { return CONSUME; }
    if is_truth(mn) { return TRUTH; }
    if i.writes_mem && !is_move(mn) { return COMMIT; }
    if is_move(mn) { return LINK; }
    ENGAGE
}

fn fields(ops: &[Op]) -> Vec<String> { ops.iter().map(|o| o.field()).collect() }

/// One instruction → its module lines. A merge emits a bare ∋ first.
fn encode(i: &x86::Insn, is_merge: bool) -> Vec<String> {
    let mut lines: Vec<String> = Vec::new();
    if is_merge { lines.push(FUSE.to_string()); }
    let mn = i.mnemonic.as_str();
    let g = classify(i);
    let f = fields(&i.ops);
    let join = |parts: &[String]| parts.join("\t");

    if g == TERM {
        lines.push(format!("{}\t{}", TERM, mn));
    } else if g == INDIRECT && matches!(mn, "syscall"|"sysenter"|"int") {
        lines.push(format!("{}\tsyscall", INDIRECT));
    } else if mn == "call" || mn == "jmp" {
        let mut parts = vec![g.to_string(), mn.to_string()]; parts.extend(f);
        lines.push(join(&parts));
    } else if mn.starts_with('j') {
        let mut parts = vec![SPLIT.to_string(), mn[1..].to_string()]; parts.extend(f);
        lines.push(join(&parts));
    } else if mn.starts_with("set") {
        let mut parts = vec![CONSUME.to_string(), mn[3..].to_string(), "set".to_string()]; parts.extend(f);
        lines.push(join(&parts));
    } else if mn.starts_with("cmov") {
        let mut parts = vec![CONSUME.to_string(), mn[4..].to_string(), "cmov".to_string()]; parts.extend(f);
        lines.push(join(&parts));
    } else {
        let glyph = if g == TRUTH { TRUTH } else if g == COMMIT { COMMIT } else if g == LINK { LINK } else { ENGAGE };
        let mut parts = vec![glyph.to_string(), mn.to_string()]; parts.extend(f);
        lines.push(join(&parts));
    }
    lines
}

/// Merge points of one function: addresses with ≥2 predecessors.
fn merges_of(insns: &[x86::Insn]) -> BTreeSet<u64> {
    let aset: BTreeSet<u64> = insns.iter().map(|i| i.addr).collect();
    let mut succ: BTreeMap<u64, u32> = BTreeMap::new();
    for (idx, i) in insns.iter().enumerate() {
        let mn = i.mnemonic.as_str();
        let falls = !(mn == "jmp" || mn.starts_with("ret"));
        if falls && idx + 1 < insns.len() { *succ.entry(insns[idx+1].addr).or_insert(0) += 1; }
        if mn.starts_with('j') {
            if let Some(t) = i.target { if aset.contains(&t) { *succ.entry(t).or_insert(0) += 1; } }
        }
    }
    succ.into_iter().filter(|&(_, c)| c >= 2).map(|(a, _)| a).collect()
}

/// ELF data sections the code reads: ALLOC, PROGBITS, not executable.
fn data_sections(raw: &[u8]) -> Vec<(u64, Vec<u8>)> {
    let mut out = Vec::new();
    if raw.len() < 64 || &raw[0..4] != b"\x7fELF" { return out; }
    let is64 = raw[4] == 2;
    if !is64 { return out; }
    let rd = |o: usize, n: usize| -> u64 { let mut v=0u64; for k in 0..n { v |= (raw[o+k] as u64) << (8*k); } v };
    let e_shoff = rd(0x28, 8) as usize;
    let e_shentsize = rd(0x3a, 2) as usize;
    let e_shnum = rd(0x3c, 2) as usize;
    for k in 0..e_shnum {
        let o = e_shoff + k * e_shentsize;
        if o + 64 > raw.len() { break; }
        let sh_type = rd(o + 4, 4);
        let sh_flags = rd(o + 8, 8);
        let sh_addr = rd(o + 16, 8);
        let sh_off = rd(o + 24, 8) as usize;
        let sh_size = rd(o + 32, 8) as usize;
        // SHT_PROGBITS(1), SHF_ALLOC(0x2), not SHF_EXECINSTR(0x4)
        if sh_type == 1 && (sh_flags & 0x2) != 0 && (sh_flags & 0x4) == 0 && sh_size > 0 {
            if sh_off + sh_size <= raw.len() {
                out.push((sh_addr, raw[sh_off..sh_off+sh_size].to_vec()));
            }
        }
    }
    out
}

/// The whole binary as an executable IMASM module. Every executable byte is
/// decoded linearly, so an address reached only through an indirect jump — a
/// switch's jump-table arm, a call through a function pointer — is in the module
/// too, not just what recursive descent reached from direct edges.
pub fn emit(raw: &[u8]) -> String {
    let (entry, segments) = crate::vox::parse_elf(raw);

    let mut out: Vec<String> = Vec::new();
    out.push(format!("; {} module", INDIRECT));
    out.push(format!("; entry 0x{:x}", entry));
    for (at, blob) in data_sections(raw) {
        let hex: String = blob.iter().map(|b| format!("{:02x}", b)).collect();
        out.push(format!("={:#x}\t{}", at, hex));
    }
    for (base, bytes) in &segments {
        let mut pos = 0usize;
        while pos < bytes.len() {
            let addr = base + pos as u64;
            match x86::decode(&bytes[pos..], addr) {
                Some(d) if d.len > 0 => {
                    out.push(format!("@0x{:x}", addr));
                    for line in encode(&d, false) { out.push(line); }
                    pos += d.len;
                }
                _ => pos += 1, // alignment padding or data between functions
            }
        }
    }
    let mut s = out.join("\n"); s.push('\n'); s
}

/// Just the structure word (glyphs), per function.
pub fn words(raw: &[u8]) -> String {
    let (entry, segments) = crate::vox::parse_elf(raw);
    let image = vox_decode::Image { segments };
    let seeds = crate::vox::elf_function_symbols(raw);
    let w = vox_decode::walk(&image, entry, &seeds);
    let mut out: Vec<String> = Vec::new();
    for (start, f) in &w.functions {
        let mut insns: Vec<x86::Insn> = Vec::new();
        for ins in f {
            if let Some(bytes) = image.bytes_at(ins.address) {
                if let Some(d) = x86::decode(bytes, ins.address) { insns.push(d); }
            }
        }
        if insns.is_empty() { continue; }
        let merges = merges_of(&insns);
        let mut word = String::from("⊢");
        for i in &insns {
            if merges.contains(&i.addr) { word.push('∋'); }
            word.push(classify(i));
        }
        out.push(format!("0x{:x}\t{}", start, word));
    }
    out.join("\n")
}

/// Function symbol names → address, from .symtab or .dynsym.
pub fn symbols(raw: &[u8]) -> BTreeMap<String, u64> {
    let mut out = BTreeMap::new();
    if raw.len() < 64 || &raw[0..4] != b"\x7fELF" || raw[4] != 2 { return out; }
    let rd = |o: usize, n: usize| -> u64 { let mut v=0u64; for k in 0..n { if o+k<raw.len(){ v |= (raw[o+k] as u64) << (8*k);} } v };
    let e_shoff = rd(0x28,8) as usize; let e_shentsize = rd(0x3a,2) as usize; let e_shnum = rd(0x3c,2) as usize;
    let sh = |k: usize, field: usize, n: usize| rd(e_shoff + k*e_shentsize + field, n);
    for k in 0..e_shnum {
        let sh_type = sh(k, 4, 4);
        if sh_type != 2 && sh_type != 11 { continue; } // SYMTAB or DYNSYM
        let sym_off = sh(k, 24, 8) as usize; let sym_size = sh(k, 32, 8) as usize;
        let strtab_idx = sh(k, 40, 4) as usize; // sh_link
        let str_off = sh(strtab_idx, 24, 8) as usize;
        let ent = sh(k, 56, 8).max(24) as usize;
        let mut o = sym_off;
        while o + 24 <= sym_off + sym_size && o + 24 <= raw.len() {
            let st_name = rd(o, 4) as usize;
            let st_info = raw.get(o+4).copied().unwrap_or(0);
            let st_value = rd(o+8, 8);
            if (st_info & 0xf) == 2 && st_value != 0 { // STT_FUNC
                let mut name = String::new();
                let mut p = str_off + st_name;
                while let Some(&b) = raw.get(p) { if b == 0 { break; } name.push(b as char); p += 1; }
                if !name.is_empty() { out.insert(name, st_value); }
            }
            o += ent;
        }
    }
    out
}
